### PR TITLE
Add support for direct method in simulator and integrate IoT Python SDK

### PR DESCRIPTION
**Pull request by avagraw**
---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to the IoT extension!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] If introducing new functionality or modified behavior, are they backed by unit and integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** unit **and** integration tests passed locally? i.e. `pytest <project root> -vv`
- [ ] Have static checks passed using the .pylintrc and .flake8 rules? Look at the CI scripts for example usage.
- [ ] Have you made an entry in HISTORY.rst which concisely explains your feature or change?

### CODE REVIEW

File: azext_iot/operations/_mqtt.py [hunk 1]
```diff
@@ -5,95 +5,76 @@
...
+        six.print_("_Payload_: {}".format(message.data))
+        if message.custom_properties:
+            six.print_("_Custom Properties_: {}".format(message.custom_properties))
+    
+    def method_request_handler(self, method_request):
+        six.print_()
+        six.print_("Received method request with id: '{}' and method name: '{}' for device with id: '{}'".format(method_request.request_id, method_request.name, self.device_id))
+        six.print_("_Payload_: {}".format(method_request.payload))
+        payload = {"result": True, "data": "Method succeeded"}  # set response payload
+        status = 200  # set return status code
```
Comment (digimaun) [inline azext_iot/operations/_mqtt.py:59]: we don't want to always hardcode 200 for the status....both the response code and response payload should have defaults, but can be customized/overridden by user.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -127,7 +127,6 @@ You can either manually set the environment variables or use the `pytest.ini.exa
     AZURE_TEST_RUN_LIVE=True
     azext_iot_testrg="Resource Group that contains your IoT Hub"
     azext_iot_testhub="IoT Hub Name"
-    azext_iot_testhub_cs="IoT Hub Connection String"
```
Comment (digimaun) [inline CONTRIBUTING.md:130]: thanks for improving these instructions

File: azext_iot/operations/_mqtt.py [hunk 3]
```diff
@@ -5,95 +5,76 @@
...
-            "Connected to target IoT Hub MQTT broker with result: {}".format(
-                connection_result[rc]
-            )
-        )
-        self.client.subscribe(self.topic_receive)
-        six.print_("Subscribed to device bound message queue")
-        self.connected = True
-
-    def on_message(self, client, userdata, msg):
+        self.registry_manager = IoTHubRegistryManager(target["cs"])
```
Comment (digimaun) [inline azext_iot/operations/_mqtt.py:19]: We shouldn't need an IoTHubRegistryManager

File: azext_iot/operations/_mqtt.py [hunk 4]
```diff
@@ -5,95 +5,76 @@
...
-
-    def on_message(self, client, userdata, msg):
+        self.registry_manager = IoTHubRegistryManager(target["cs"]) 
+        self.device_client = mqtt_device_client.create_from_connection_string(device_connection_string)
+        self.digital_twin_client =  DigitalTwinClient(target["cs"])
+        self.device_client.connect()
+        self.device_client.on_message_received = self.message_handler
+        self.device_client.on_method_request_received = self.method_request_handler
+
+    def send_c2d_message(self, message_text="", properties={}):
```
Comment (digimaun) [inline azext_iot/operations/_mqtt.py:26]: send_c2d_message is out of scope for this PR and should be removed.

File: azext_iot/operations/_mqtt.py [hunk 5]
```diff
@@ -5,95 +5,76 @@
...
+            result.status = "Failed to send C2D Message"
+            result.error = "Unexpected error {0}".format(ex)
+        return result
+    
+    def send_d2c_message(self, message_text="", properties={}):
+        message = Message(message_text)
+        message.custom_properties = properties
+        self.device_client.send_message(message)
+    
+    def invoke_direct_method(self, command_name, payload="", connect_timeout_in_seconds=100, response_timeout_in_seconds=100):
```
Comment (digimaun) [inline azext_iot/operations/_mqtt.py:44]: invoke_direct_method here is out of scope for this PR and should be removed.

**Comment by digimaun** [comment]
Was expecting updated dependency references in `setup.py`

File: azext_iot/operations/hub.py [hunk 6]
```diff
@@ -2206,14 +2206,13 @@ def http_wrap(target, device_id, generator):
...
-            wrap.execute(generator(), publish_delay=msg_interval, msg_count=msg_count)
+            device_connection = iot_get_device_connection_string(cmd=cmd, device_id=device_id, hub_name=hub_name, login=login)
+            client_mqtt = mqtt_client(
+                            target=target,
+                            device_connection_string=device_connection["connectionString"],
+                            device_id=device_id
+                        )
+            client_mqtt.execute(data=generator(), properties=properties_to_send, publish_delay=msg_interval, msg_count=msg_count)
         else:
             six.print_("Sending and receiving events via https")
```
Comment (digimaun) [inline azext_iot/operations/hub.py:2217]: Does the python device client also support http? If so we need to align the http behavior as well.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -264,7 +263,7 @@ https://medium.com/@marcobelo/setting-up-python-black-on-visual-studio-code-5318
 
 https://docs.python.org/3/library/pdb.html
 
-1. `pip install pdb`
```
[the remark this answers is not in the document]
Comment (avagraw) [reply CONTRIBUTING.md:267]: so pdb has been replaces by pdb++ and I feel its good for completeness to include this

File: azext_iot/operations/hub.py (hunk 6 above)
Comment (avagraw) [reply azext_iot/operations/hub.py:2217]: HTTP is not supported by the Python device client

File: azext_iot/tests/iothub/test_iot_messaging_int.py [hunk 8]
```diff
@@ -217,7 +259,11 @@ def test_uamqp_device_messaging(self):
             expect_failure=True,
         )
 
-    def test_device_messaging(self):
+    @pytest.mark.skipif(
+        not validate_min_python_version(3, 4, exit_on_fail=False),
```
Comment (digimaun) [inline azext_iot/tests/iothub/test_iot_messaging_int.py:263]: The minimum python we support now is 3.6 so we don't need this skipif mark (or others targeting 3.5 and below).

File: azext_iot/_params.py [hunk 9]
```diff
@@ -608,6 +608,16 @@ def load_arguments(self, _):
             options_list=["--properties", "--props", "-p"],
             help=info_param_properties_device(include_http=True),
         )
+        context.argument(
+            "method_response_status_code",
+            options_list=["--method-response-status-code", "--mrsc"],
```
Comment (digimaun) [inline azext_iot/_params.py:613]: I think we can shorten this to `--method-reponse-code`

File: setup.py [hunk 10]
```diff
@@ -48,8 +48,7 @@
 # though that is installed out of band (managed by the extension)
 # for compatibility reasons.
 
-DEPENDENCIES = ["paho-mqtt==1.5.0", "jsonschema==3.2.0", "packaging"]
-
+DEPENDENCIES = ["paho-mqtt>=1.5.0", "jsonschema==3.0.2", "packaging", "azure-iot-device>=2.5.1"]
```
Comment (digimaun) [inline setup.py:51]: if we add `azure-iot-device` here we should remove it from dev_requirements

Comment (digimaun) [reply setup.py:51]: IIRC `urllib3[secure]>=1.21.1,<=1.25` in dev_requirements was specifically for `azure-iot-device` so it should be removed from there as well.
